In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/apple-quality/apple_quality.csv


# **Import Dataset**

In [2]:
df = pd.read_csv('/kaggle/input/apple-quality/apple_quality.csv')
df

A_id      Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0        0.0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1        1.0 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2        2.0 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3        3.0 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4        4.0  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   
...      ...       ...       ...        ...          ...        ...       ...   
3996  3996.0 -0.293118  1.949253  -0.204020    -0.640196   0.024523 -1.087900   
3997  3997.0 -2.634515 -2.138247  -2.440461     0.657223   2.199709  4.763859   
3998  3998.0 -4.008004 -1.779337   2.366397    -0.200329   2.161435  0.214488   
3999  3999.0  0.278540 -1.715505   0.121217    -1.154075   1.266677 -0.776571   
4000     NaN       NaN       NaN        NaN          NaN        NaN       NaN   

                                Acidity Quality  
0                          -0.491590483    good  
1                          -0.722809367    good  
2                           2.621636473     bad  
3                           0.790723217    good  
4                           0.501984036    good  
...                                 ...     ...  
3996                        1.854235285    good  
3997                       -1.334611391     bad  
3998                       -2.229719806    good  
3999                        1.599796456    good  
4000  Created_by_Nidula_Elgiriyewithana     NaN  

[4001 rows x 9 columns]

# **Basic Checks**

In [3]:
# first five rows
df.head()

A_id      Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0   0.0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1   1.0 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2   2.0 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3   3.0 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4   4.0  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   

        Acidity Quality  
0  -0.491590483    good  
1  -0.722809367    good  
2   2.621636473     bad  
3   0.790723217    good  
4   0.501984036    good

In [4]:
# last five rows
df.tail()

A_id      Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
3996  3996.0 -0.293118  1.949253  -0.204020    -0.640196   0.024523 -1.087900   
3997  3997.0 -2.634515 -2.138247  -2.440461     0.657223   2.199709  4.763859   
3998  3998.0 -4.008004 -1.779337   2.366397    -0.200329   2.161435  0.214488   
3999  3999.0  0.278540 -1.715505   0.121217    -1.154075   1.266677 -0.776571   
4000     NaN       NaN       NaN        NaN          NaN        NaN       NaN   

                                Acidity Quality  
3996                        1.854235285    good  
3997                       -1.334611391     bad  
3998                       -2.229719806    good  
3999                        1.599796456    good  
4000  Created_by_Nidula_Elgiriyewithana     NaN

In [5]:
# shape of the dataset
df.shape

(4001, 9)

In [6]:
# info abou the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4001 entries, 0 to 4000
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   A_id         4000 non-null   float64
 1   Size         4000 non-null   float64
 2   Weight       4000 non-null   float64
 3   Sweetness    4000 non-null   float64
 4   Crunchiness  4000 non-null   float64
 5   Juiciness    4000 non-null   float64
 6   Ripeness     4000 non-null   float64
 7   Acidity      4001 non-null   object 
 8   Quality      4000 non-null   object 
dtypes: float64(7), object(2)
memory usage: 281.4+ KB


# **Data Preprocessing**

In [7]:
# checking for the null values
df.isnull().sum()

A_id           1
Size           1
Weight         1
Sweetness      1
Crunchiness    1
Juiciness      1
Ripeness       1
Acidity        0
Quality        1
dtype: int64

It's Shows that only in one rows almost all values are missing so we have to drop it

In [8]:
# Drop rows with null values
df = df.dropna()

In [9]:
df.isnull().sum()

A_id           0
Size           0
Weight         0
Sweetness      0
Crunchiness    0
Juiciness      0
Ripeness       0
Acidity        0
Quality        0
dtype: int64

In [10]:
df.shape

(4000, 9)

In [11]:
# checking for duplicated values

df.duplicated().sum()

0

In [12]:
df.head()

A_id      Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0   0.0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1   1.0 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2   2.0 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3   3.0 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4   4.0  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   

        Acidity Quality  
0  -0.491590483    good  
1  -0.722809367    good  
2   2.621636473     bad  
3   0.790723217    good  
4   0.501984036    good

In [13]:
# We eliminate the first id column since it does not add value to the analysis.
df = df.drop("A_id", axis = 1)
df.head()

Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   

        Acidity Quality  
0  -0.491590483    good  
1  -0.722809367    good  
2   2.621636473     bad  
3   0.790723217    good  
4   0.501984036    good

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4000 entries, 0 to 3999
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Size         4000 non-null   float64
 1   Weight       4000 non-null   float64
 2   Sweetness    4000 non-null   float64
 3   Crunchiness  4000 non-null   float64
 4   Juiciness    4000 non-null   float64
 5   Ripeness     4000 non-null   float64
 6   Acidity      4000 non-null   object 
 7   Quality      4000 non-null   object 
dtypes: float64(6), object(2)
memory usage: 281.2+ KB


In [15]:
# Acidity shows object dtype so we have to change that into float 

df['Acidity'] = df['Acidity'].astype('float64')

## Handling Categorical Data

In [16]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
df['Quality'] = label_encoder.fit_transform(df['Quality'])

In [17]:
df.head()

Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness   Acidity  \
0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840 -0.491590   
1 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530 -0.722809   
2 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033  2.621636   
3 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761  0.790723   
4  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849  0.501984   

   Quality  
0        1  
1        1  
2        0  
3        1  
4        1

# Scaling the Numerical Columns

In [18]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
df[['Size', 'Weight', 'Sweetness', 'Crunchiness', 'Juiciness', 'Ripeness', 'Acidity']] = scaler.fit_transform(df[['Size', 'Weight', 'Sweetness', 'Crunchiness', 'Juiciness', 'Ripeness', 'Acidity']])

In [19]:
df

Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0    -1.798424 -0.950373   2.993421    -1.424150   0.690545 -0.089872   
1    -0.359060 -1.154404   2.127698     0.429746   0.176767  0.197020   
2     0.109445 -0.225759  -0.652507    -0.946892   1.205422 -0.286156   
3    -0.079977 -0.800146   0.923916    -0.772399   1.619575 -2.087320   
4     0.968573 -0.191640   0.044164    -1.096894   1.305025 -0.961548   
...        ...       ...        ...          ...        ...       ...   
3995  0.291729 -0.048594  -1.669449    -0.365345   0.614425  0.931482   
3996  0.108878  1.834105   0.137124    -1.159058  -0.252634 -0.846326   
3997 -1.105655 -0.716904  -1.013784    -0.234036   0.874379  2.275957   
3998 -1.818112 -0.492908   1.459901    -0.845446   0.854549 -0.151419   
3999  0.405409 -0.453071   0.304496    -1.525439   0.390954 -0.680212   

       Acidity  Quality  
0    -0.269415        1  
1    -0.378997        1  
2     1.206044        0  
3     0.338315        1  
4     0.201472        1  
...        ...      ...  
3995  0.028866        0  
3996  0.842347        1  
3997 -0.668950        0  
3998 -1.093171        1  
3999  0.721761        1  

[4000 rows x 8 columns]

## Splitting Data into X and y

In [20]:
X = df.drop('Quality', axis =1)
y = df['Quality']

# **Model Building**

In [21]:
# Importing some needed Libararies

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, f1_score, classification_report
from sklearn.neural_network import MLPClassifier

In [22]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [23]:
# Listof classifier to try

classifiers = {
     'Logistic Regression': LogisticRegression(),
    'Decision tree': DecisionTreeClassifier(),
    'Random Forest': RandomForestClassifier(),
    'SVM': SVC(),
    'K-nearest Neigbour': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(),
    'XGBoost':XGBClassifier()
    
}

In [24]:
# List to store accuarcy and f1 score
accuracies = []
f1_scores = []

In [25]:
# train and ecvaluate each classifer
for name, clf in classifiers.items():
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    
    accuracy = accuracy_score(y_test, y_pred)
    
    accuracies.append(accuracy)
    
    print(f"\n{30*'='}\n{name}\n{30*'='}")
    print(f"Accuracy: {accuracy:.4f}")
    print("Classification Report:")
    print(classification_report(y_test, y_pred))


Logistic Regression
Accuracy: 0.7538
Classification Report:
              precision    recall  f1-score   support

           0       0.75      0.76      0.75       401
           1       0.75      0.75      0.75       399

    accuracy                           0.75       800
   macro avg       0.75      0.75      0.75       800
weighted avg       0.75      0.75      0.75       800


Decision tree
Accuracy: 0.8037
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.81      0.81       401
           1       0.81      0.80      0.80       399

    accuracy                           0.80       800
   macro avg       0.80      0.80      0.80       800
weighted avg       0.80      0.80      0.80       800


Random Forest
Accuracy: 0.9000
Classification Report:
              precision    recall  f1-score   support

           0       0.90      0.90      0.90       401
           1       0.90      0.90      0.90       399

    accurac

/opt/conda/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(



Neural Network
Accuracy: 0.9375
Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.95      0.94       401
           1       0.95      0.93      0.94       399

    accuracy                           0.94       800
   macro avg       0.94      0.94      0.94       800
weighted avg       0.94      0.94      0.94       800


XGBoost
Accuracy: 0.9062
Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.91      0.91       401
           1       0.91      0.91      0.91       399

    accuracy                           0.91       800
   macro avg       0.91      0.91      0.91       800
weighted avg       0.91      0.91      0.91       800



## Classifier Performance Report

This report provides an overview of the performance of various classifiers on the task of predicting the quality of apples based on multiple attributes. The classifiers evaluated include Logistic Regression, Decision Tree, Random Forest, SVM, K-Nearest Neighbors, Neural Network, and XGBoost.

### 1. Logistic Regression:
- **Accuracy:** 75.38%
- **F1 Score:** 0.75
- **Summary:** Logistic Regression achieved a balanced accuracy with equal precision, recall, and F1 score for both classes (good and bad).

### 2. Decision Tree:
- **Accuracy:** 80.87%
- **F1 Score:** 0.81
- **Summary:** The Decision Tree classifier demonstrated good accuracy and balanced performance for both classes, making it a robust choice.

### 3. Random Forest:
- **Accuracy:** 90.12%
- **F1 Score:** 0.90
- **Summary:** Random Forest exhibited outstanding accuracy and F1 score, suggesting its ability to capture complex patterns in the data.

### 4. SVM:
- **Accuracy:** 90.38%
- **F1 Score:** 0.90
- **Summary:** The SVM classifier provided high accuracy and F1 score, indicating its effectiveness in handling non-linear relationships.

### 5. K-Nearest Neighbors:
- **Accuracy:** 90.12%
- **F1 Score:** 0.90
- **Summary:** K-Nearest Neighbors demonstrated strong performance, aligning with other high-performing classifiers.

### 6. Neural Network:
- **Accuracy:** 91.75%
- **F1 Score:** 0.92
- **Summary:** The Neural Network exhibited superior accuracy and F1 score, showcasing its ability to model complex relationships within the data.

### 7. XGBoost:
- **Accuracy:** 90.62%
- **F1 Score:** 0.91
- **Summary:** XGBoost provided excellent accuracy and F1 score, making it a powerful ensemble learning technique.

### Overall Observations:
- The models achieved high accuracy and F1 scores, indicating their capability to effectively classify apples into quality categories.
- Neural Network and Random Forest outperformed other models, showcasing their ability to handle intricate patterns in the data.
- It's important to note that the Neural Network showed a convergence warning, suggesting that increasing the number of iterations may further improve its performance.


